dtm0/dtx.c

```diff
@@ -336,68 +343,90 @@ static void dtx_exec_all_ast_cb(struct m0_sm_group *grp, struct m0_sm_ast *ast)
+	rec = m0_be_dtm0_log_find(log, &txd->dtd_id);
+
+	if (rec != NULL) {
+		dtx = &rec->dlr_dtx;
+		m0_dtm0_tx_desc_apply(&dtx->dd_txd, txd);
+		dtx_log_update(dtx);
+
+		if (dtx->dd_sm.sm_state == M0_DDS_EXECUTED_ALL &&
+		    m0_dtm0_tx_desc_state_exists(&dtx->dd_txd,
+						 M0_DTPS_PERSISTENT)) {
+			M0_ASSERT(dtx->dd_nr_executed ==
+				  dtx->dd_txd.dtd_ps.dtp_nr);
+			M0_LOG(M0_DEBUG, "dtx " DTID0_F "is stable (PMA)",
+			       DTID0_P(&dtx->dd_txd.dtd_id));
+			m0_sm_state_set(&dtx->dd_sm, M0_DDS_STABLE);
+		}
 	}
+	m0_mutex_unlock(&log->dl_lock);
 
-out:
 	m0_free(pma);
 	m0_fop_put_lock(fop); /* it was taken in ::m0_dtm0_dtx_post_pmsg */
 	M0_LEAVE();
 }
 
-M0_INTERNAL void m0_dtm0_dtx_pmsg_post(struct m0_dtm0_dtx *dtx,
-				       struct m0_fop      *fop)
+M0_INTERNAL void m0_dtm0_dtx_pmsg_post(struct m0_be_dtm0_log *log,
+				       struct m0_fop         *fop)
 {
-	struct m0_dtm0_pmsg_ast *pma;
+	struct m0_dtm0_pmsg_ast      *pma;
+	struct m0_sm_group           *dtx_sm_grp;
+	struct m0_dtm0_log_rec       *rec;
+	struct dtm0_req_fop          *req = m0_fop_data(fop);
+	const struct m0_dtm0_tx_desc *txd = &req->dtr_txr;
 
 	M0_PRE(fop != NULL);
 	M0_PRE(fop->f_opaque != NULL);
 	M0_PRE(m0_fop_data(fop) != NULL);
-	M0_PRE(dtx->dd_sm.sm_grp != &fop->f_item.ri_rmachine->rm_sm_grp);
+	M0_PRE(fop->f_type == &dtm0_req_fop_fopt);
+	M0_PRE(!log->dl_is_persistent);
+
+	M0_ENTRY("fop=%p", fop);
 
-	M0_ENTRY("dtx=%p, fop=%p", dtx, fop);
 	/*
-	 * XXX: it is an uncommon case where f_opaque is used on the server
-	 * side (rather than on the client side) to associate a FOP with
-	 * a user-defined datum. It helps to connect the lifetime of the
-	 * AST that handles the notice with the lifetime of the FOP that
-	 * has delivered the notice.
+	 * This lookup is done to find the SM group that corresponds
+	 * to DTX SM to post AST. We can not use DTX itself here since
+	 * it can be cleared in another thread and the log lock does
+	 * not protect the DTX between current function and AST.
 	 */
-	pma = fop->f_opaque;
-	M0_ASSERT_INFO(M0_IS0(pma), "PMA cannot be re-used");
-	*pma = (struct  m0_dtm0_pmsg_ast) {
-		.p_dtx = dtx,
-		.p_fop = fop,
-		.p_ast = {
-			.sa_cb = dtx_persistent_ast_cb,
-			.sa_datum = pma,
-		},
-	};
-	m0_fop_get(fop); /* it will be put back by ::dtx_persistent_ast_cb */
-	m0_sm_ast_post(dtx->dd_sm.sm_grp, &pma->p_ast);
+	m0_mutex_lock(&log->dl_lock);
+	rec = m0_be_dtm0_log_find(log, &txd->dtd_id);
+	dtx_sm_grp = rec != NULL ? rec->dlr_dtx.dd_sm.sm_grp : NULL;
+	m0_mutex_unlock(&log->dl_lock);
+
+	if (rec != NULL) {
+		/*
+		 * XXX: it is an uncommon case where f_opaque is used on the
+		 * server side (rather than on the client side) to associate
+		 * a FOP with a user-defined datum. It helps to connect the
+		 * lifetime of the AST that handles the notice with the lifetime
+		 * of the FOP that has delivered the notice.
+		 */
+		pma = fop->f_opaque;
+		M0_ASSERT_INFO(M0_IS0(pma), "PMA cannot be re-used");
```

Comment:
It's an expensive check. Better to set it to NULL after it's freed at `dtx_persistent_ast_cb()`, like @huanghua78 suggested, to make sure it's not re-used.

Reply:
Done.